src/mmds/src/token.rs

```diff
@@ -0,0 +1,367 @@
+            .expect("Failed to compute token expiry date.")
+    }
+
+    /// Specifies whether the token has expired.
+    pub fn has_expired(&self) -> bool {
+        Utc::now()
+            .signed_duration_since(self.expiry())
+            .num_milliseconds()
+            >= 0
+    }
+
+    /// Validate the token time to live against bounds.
+    pub fn check_ttl(ttl: u32) -> bool {
+        (MIN_TOKEN_TTL_SECONDS..=MAX_TOKEN_TTL_SECONDS).contains(&ttl)
+    }
+}
+
+/// Structure for the Token used by IMDSv2.
+#[derive(Clone)]
+pub struct Token {
+    value: String,
+    token_info: TokenInfo,
+}
+
+impl Token {
+    /// Create a new session Token for MMDS requests.
+    fn new(value: String, token_info: TokenInfo) -> Self {
+        Token { value, token_info }
+    }
+
+    /// Generate a new random token with the specified lifetime.
+    pub fn generate(ttl_seconds: u32) -> Result<Self, TokenError> {
+        // Validate token time to live.
+        if !TokenInfo::check_ttl(ttl_seconds) {
+            return Err(TokenError::InvalidTtlValue(ttl_seconds));
+        }
+
+        let mut random_bytes = [0u8];
+        let mut token: String = String::new();
+        let mut random_file =
+            File::open(Path::new(RANDOMNESS_POOL)).map_err(TokenError::EntropyPool)?;
+
+        while token.len() < TOKEN_LEN {
+            random_file
+                .read(&mut random_bytes)
+                .map_err(TokenError::ReadFromEntropyPool)?;
+            // Token only accepts alphanumeric characters.
+            if !random_bytes[0].is_ascii_alphanumeric() {
+                continue;
+            }
+            if let Ok(str) = String::from_utf8(random_bytes.to_vec()) {
+                token.push_str(&*str);
+            }
+        }
+
+        Ok(Self::new(token, TokenInfo::new(ttl_seconds)))
+    }
+
+    /// Specifies whether the token has expired.
+    pub fn has_expired(&self) -> bool {
+        self.token_info.has_expired()
+    }
+
+    /// Get token string sequence.
+    pub fn value(&self) -> &String {
+        &self.value
+    }
+}
+
+pub struct TokenStore {
+    tokens: HashMap<String, TokenInfo>,
+}
+
+impl Default for TokenStore {
+    fn default() -> Self {
+        TokenStore {
+            tokens: HashMap::default(),
+        }
+    }
+}
+
+impl TokenStore {
+    /// Generate a new token and add it to the store.
+    pub fn generate(&mut self, ttl_seconds: u32) -> Result<Token, TokenStoreError> {
+        // Removed expired tokens before adding a new one.
+        self.remove_expired();
+
+        // Verify that maximum number of session tokens has not been reached.
+        if self.tokens.len() >= MAX_SESSIONS {
+            return Err(TokenStoreError::TokensLimitExceeded);
+        }
+
+        let token = Token::generate(ttl_seconds).map_err(TokenStoreError::TokenGenerationError)?;
+        let result = token.clone();
+        self.tokens.insert(token.value, token.token_info);
+        Ok(result)
```

Comment:
It doesn't seem that `TokenInfo/TokenTTL` has any meaning or use outside this module; actually outside of `struct TokenStore`.

All users validate using `pub fn is_valid(&mut self, token_value: String) -> bool` so they only need the value, not the ttl. Ttl is only used internally right?

If that's the case, I would simplify this file a bit by dropping `Token` object (altogether or at least from public interface). In public interface all you need is:
- `generate_token(ttl_seconds) -> String` - generates `token_value: String` and also adds private `TokenTTL` to hashmap for that string.
- `is_valid(&mut self, token_value: String) -> bool` - exactly as it is now ; checks TTL in hashmap at key `token_value`

Reply:
I think keeping `Token` makes for a more clear implementation. However, I dropped it from the public interface by making `generate_token()` return `String` instead of `Token` object. Let me know what you think.